src/site/content/en/reliable/two-way-communication-guide/index.md

```diff
@@ -0,0 +1,400 @@
+browsers](https://caniuse.com/?search=channel).
+
+### Advanced APIs: Background Sync and Background Fetch
+
+In this guide we explored ways of implementing **two-way** communication techniques, for relatively
+simple cases, like passing a string message describing the operation to perform, or a list of URLs
+to cache from one context to the other. In this section we'll explore two APIs to handle specific
+scenarios: lack of connectivity and long downloads.
+
+#### Background Sync
+
+A chat app might want to make sure that messages are never lost due to bad connectivity. The
+[Background Sync API](https://developers.google.com/web/updates/2015/12/background-sync) lets you
+defer actions to be retried when the user has stable connectivity. This is useful for ensuring that
+whatever the user wants to send, is actually sent. 
+
+Instead of the `postMessage()` interface, the page register for a `sync`:
+
+```javascript
+navigator.serviceWorker.ready.then(function (swRegistration) {
+  return swRegistration.sync.register('myFirstSync');
+});
+```
+
+The service worker then listens for the `sync` event to process the message:
+
+```javascript
+self.addEventListener('sync', function (event) {
+  if (event.tag == 'myFirstSync') {
+    event.waitUntil(doSomeStuff());
+  }
+});
+```
+
+The function `doSomeStuff()` should return a promise indicating the success/failure of whatever it's
+trying to do. If it fulfills, the sync is complete. If it fails, another sync will be scheduled to
+retry. Retry syncs also wait for connectivity, and employ an exponential back-off. 
+
+Once the operation has been performed, the service worker can then communicate back with the page to
+update the UI, by using any of the communication APIs explored earlier.
+
+Google search uses `Background Sync` to persist failed queries due to bad connectivity, and retry
+them later when the user is online. Once the operation is performed, they communicate the result to
+the user via a web push notification:
+
+<figure class="w-figure">
+  <img src="google-search-background-sync.png"
+       width="700"
+       alt="Diagram showing a page passing a port to a service worker, to establish two-way communication.">
+</figure>
+
+{% Aside %} Check out [this guide](https://web.dev/resilient-search-experiences/) to see this
+feature in your site using [Workbox Background
+Sync](https://developers.google.com/web/tools/workbox/modules/workbox-background-sync). {% endAside
+%}
+
+#### Background Fetch
+
+For relatively short bits of work like sending a message, or a list of URLs to cache, the options
+explored so far are a good choice. If the task takes too long the browser will kill the service
+worker, otherwise it's a risk to the user's privacy and battery. 
+
+The [Background Fetch API](https://developers.google.com/web/updates/2018/12/background-fetch)
+allows you to offload a long task to a service worker, like downloading a movie, podcasts, or levels
```

Comment:
```suggestion
allows you to offload a long task to a service worker, like downloading movies, podcasts, or levels
```
https://owl.purdue.edu/owl/general_writing/mechanics/parallel_structure.html#:~:text=Parallel%20structure%20means%20using%20the,and%22%20or%20%22or.%22